builder.md

```diff
@@ -1,57 +1,142 @@
+### CLI
 
-In the case where schemes share the same slug, a builder will overwrite files previously generated from the template. Should this happen, a builder should show warning messages listing the overwritten files.
+It is REQUIRED that that this functionality is exposed as binary CLI executable
+with the exact interface described below.
+
+The binary name SHOULD contain `base16`, but is otherwise left as a choice to
+the author.
+
+To be fully compliant, a builder CLI interface MUST NOT include any other
+feature, option, argument, or deviance from the expected interface and
+behaviour of the program.
+
+<!-- TODO: For convenience, we make a manpage and set of tests available. All
+compliant builders MUST fully conform to these two. -->
+
+The CLI interface MUST work without relying on any network connection.
+
+```bash
+base16 TEMPLATES-DIRECTORY SCHEME-FILE ...
+```
+
+`TEMPLATES-DIRETORY` being a directory containing a `config.yaml`, as well as at
+least one `.mustache` template. These are usually named `templates` on all
+base16 template repositories.
+
+`SCHEME-FILE` being one or more `.yaml` scheme file. The main source for these
+is the [base16-schemes](https://github.com/base16-project/base16-schemes)
+repository.
+
+The program MAY offer a single `--debug | -d` option for increasing log
+verbosity. These are considered implementation detail and each author SHOULD
+implement them as they see fit.
+
+### Library
+
+The compliant base16 builder software MAY also expose a software library other
+developers may use to assist developing more complex base16-compatible
+software.
+
+This exposed library, or any internal code, has no required structure or usage.
+
+The author MAY choose how they will expose these functionalities to the caller,
+according to their preferences and SHOULD follow best practices on their
+respective programming languages.
+
+It is RECOMMENDED that builders follow semantic versioning for their library
+interface.
+
+The author MAY implement additional features that are exposed through the
+library, as long as it does not affect the CLI functionality compliance.
+
+## Output and behaviour
+
+**Note**: As the CLI is not intended for usual human usage, all outputted text
+messages are considered implementation detail, so the author MAY output
+whatever they prefer (or no message at all). If needed, scripts using the
+builder SHOULD check for return codes (specified below) instead of messages.
+
+For all templates defined in the template config file (`config.yaml`, inside
+the specified template directory), the builder MUST iterate through all the
+defined schemes and output matching files.
+
+The built filename should look like [output-dir]/base16-[slug][extension],
+where the slug is taken from the scheme filename made lowercase with spaces
+replaced with dashes and both the extension and output-dir are taken from
+`config.yaml`.
+
+The builder MUST check for the (unusual) case where schemes share the same
```

Comment:
Couldn't (shouldn't?) we test for this in the schemes repo? Have some tests running and include this case? I suppose we should still check for this here because someone could clone and edit the schemes dir while testing out a colourscheme and give duplicate a slug value.

Reply:
Yeah, we shouldn't any schemes with colliding slugs (actually, all file names on our repo should already be slugified).

And yup, that's exactly it. It could be useful to error out when someone with a modified scheme collection accidentally two with the same slug, which would be very hard to debug.

My idea is basically  erroring out only when the _inputs_ are the same, the outputs should be overwritten silently; as it is a common usecase to run the command to rebuild when the output already exists, and current builders spam a lot of warnings that are only useful to debug that rare collision problem.